### ISSUE TITLE
Extremely slow rendering of the TTF version FuraCode font from the Nerd Fonts project

### ISSUE THREAD

**Issue by kailiu42** [snowie2000/mactype#513]
I'm using the [FuraCode](https://github.com/ryanoasis/nerd-fonts/tree/master/patched-fonts/FiraCode) font from the [Nerd Fonts](https://github.com/ryanoasis/nerd-fonts) project.

The problem is the TTF version is rendered really slow, while the OTF version is rendered normally. When using the TTF version, it took almost a second for the text on a PuTTY terminal window to display, progressively.

The problem might not be solely on MacType side, since if I exclude PuTTY I still can notice the text display is a bit lower than normal. But it is still acceptable. With MacType rendering, the slowness became so obvious that is almost not usable. Imaging a game running at 10FPS, that's the feeling.

OS: Windows 10 1809.
MacType: 2018.1-beta5

**Comment by snowie2000** [comment]
Alright, turns out that the FuraCode font has many different weights and several ttf files for each weight.
Which specific version are you using?

**Comment by kailiu42** [comment]
I'm using the "Mono Windows Compatible" fonts. Specifically these files:

Fura Code Bold Nerd Font Complete Mono Windows Compatible.ttf
Fura Code Light Nerd Font Complete Mono Windows Compatible.ttf
Fura Code Medium Nerd Font Complete Mono Windows Compatible.ttf
Fura Code Regular Nerd Font Complete Mono Windows Compatible.ttf

**Comment by snowie2000** [comment]
I haven't tested it in Windows10, but the font works very well in Windows 7.
This is a screen recording from my Windows 7 system with Fira Code.

[putty.zip](https://github.com/snowie2000/mactype/files/2898675/putty.zip)

**Comment by kailiu42** [comment]
I noticed the font name in the video and your comment is Fira Code. Did you tested the original Fira Code or the Fura Code from Nerd Fonts?

**Comment by snowie2000** [comment]
I didn't see the FuraCode NF fonts because their names turned into garbled texts when MacType is enabled. I thought the name after installation is actually FiraCode 🤣

Checking the problem.

**Comment by snowie2000** [comment]
![qq 20190225164555](https://user-images.githubusercontent.com/11767189/53324868-db24cb80-391c-11e9-8493-5a6ec9de2fb2.png)
There is one problem: 
Even when I turned off the MacType and re-launch putty, the FuraCode NF fonts still refuse to show up.
It is available in the list of Notepad and other applications.

**Comment by kailiu42** [comment]
That's strange. I'm using the latest snapshot of PuTTY and I just tried the stable release 0.70, there is no problem to see FuraCode NF in the font dialog.

**Comment by kailiu42** [comment]
Some latest test feedback:

- The original Fira Code TTF font seems don't have such slow rendering issue.
- When I already have FuraCode NF installed, then installed the original Fira Code TTF fonts, the font selection dialog of PuTTY and NotePad cannot correctly show Fira Code in the font list, but some other incorrect name. However if selected that font it shows Fira Code in the above text edit control. See below screenshot. Maybe this is the similar problem as yours?

![snipaste_2019-02-26_20-43-14](https://user-images.githubusercontent.com/358679/53413675-92960c80-3a07-11e9-9f74-69def3bfef25.png)

**Comment by snowie2000** [comment]
I noticed this problem on my first try and till now I finally got some time digging into it.

Turns out that the problem is from the font `FuraCode NF`, it incorrectly has the same `UniqueID` as the `Fira Code`: 1.206;CTDB;FiraCode-Light
![qq 20190228155415](https://user-images.githubusercontent.com/11767189/53550323-60082300-3b71-11e9-8225-b3d5a720a4df.png)

Windows APIs use the `UniqueID` field as the full name of a font, and MacType distinguishes fonts with it. If two fonts have the same ID, they will be treated as a single font, thus drawing text with glyph index will result in garbled text.

**Comment by snowie2000** [comment]
Just pushed a commit to improve the ability to distinguish fonts. 

Now the font dialog looks normal but still got no luck to make the FuraCode NF show up in the putty.
 
![qq 20190228165147](https://user-images.githubusercontent.com/11767189/53554239-428b8700-3b7a-11e9-8035-3f9766bd6ccb.png)

**Comment by kailiu42** [comment]
On my system, the slowness it not solely on PuTTY, any program using the TTF FuraCode NF are slower than OTF. So maybe you can just try the NotePad, use it to open a large text file and scroll around to see the symptom.

**Comment by kailiu42** [comment]
BTW my screen is 4K, does it make MacType rendering considerably slower?

**Comment by sammilucia** [comment]
On my computer MacType renders a bit slower than GDI or DirectWrite, but you would never notice. It is possible FuraCode TTF is just really poorly optimised (e.g. it draws font stems several times or has extremely complex paths).

**Comment by snowie2000** [comment]
I didn't notice any slowing down when using FuraCode NF or Fira Code. 

Although 4K can definitely slow things down, it shouldn't be that noticeable.

**Comment by snowie2000** [comment]
[MacType-fura-special.zip](https://github.com/snowie2000/mactype/files/2917538/MacType-fura-special.zip)

Try this version. At least, it can correctly show the names of the two fonts.

**Comment by kailiu42** [comment]
Yes the font name issue is gone.

**Comment by sammilucia** [comment]
@kraml is the slowness problem still occurring?

**Comment by kailiu42** [comment]
Yes it is still slow. The latest build provided above only solved the font name issue.

**Comment by sammilucia** [comment]
@kraml okay thank you.
